node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +61,379 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
+					OverseerFieldInit<#baggage_ty>
+				},
+				uninited_ty: quote! {
+					OverseerFieldUninit<#baggage_ty>
+				},
+				order: idx + last_subsystem_idx
+			}
+		})
+	);
+
+	let struct_common_generics = field_helpers
+		.iter()
+		.filter(|fh| fh.is_generic)
+		.map(|fh| &fh.field_ty)
+		.collect::<Vec<_>>();
+	let builder_common_generics = field_helpers
+		.iter()
+		.map(|fh| &fh.placeholder_type)
+		.collect::<Vec<_>>();
+	let where_clause_elts = field_helpers
+		.iter()
+		.filter_map(|fh| fh.where_clause.as_ref())
+		.collect::<Vec<_>>();
+
+	let builder_where_clause = quote! {
 		where
-			S: #support_crate ::SpawnNamed,
+			#( #where_clause_elts, )*
+			S: #support_crate ::SpawnNamed
 	};
 
-	let builder_generics = quote! {
-		<S, #( #baggage_generic_ty, )* #( #builder_generic_ty, )* >
-	};
+	let builder_subsystem_setters = field_helpers
+		.iter()
+		.enumerate()
+		.map(|(idx, field_helper)| {
+			let setter_name = &field_helper.name;
+			let var_type = &field_helper.field_ty;
+
+			let impl_generics = if !field_helper.is_generic {
+				// For non generics we exclude type definition for the field being set
+				field_helpers
+					.iter()
+					.filter(|other_field| other_field.order != idx)
+					.map(|other_field| &other_field.placeholder_type)
+					.collect::<Vec<_>>()
+			}
+			else {
+				// For generics we replace placeholder type with a real generic type
+				field_helpers
+					.iter()
+					.map(|other_field| {
+						if other_field.order == idx {
+							&other_field.field_ty
+						}
+						else {
+							&other_field.placeholder_type
+						}
+					})
+					.collect::<Vec<_>>()
+			};
+			// For the type generics we replace placeholder type with Uninit<T>
+			let type_specific_generics = field_helpers
+				.iter()
+				.map(|other_field| {
+					if other_field.order == idx {
+						&other_field.uninited_ty
+					}
+					else {
+						&other_field.placeholder_type
+					}
+				})
+				.collect::<Vec<_>>();
+			// For the setter return type generics we replace placeholder type with Init<T>
+			let return_type_generics = field_helpers
+				.iter()
+				.map(|other_field| {
+					if other_field.order == idx {
+						&other_field.inited_ty
+					}
+					else {
+						&other_field.placeholder_type
+					}
+				})
+				.collect::<Vec<_>>();
+			// To set specific fields in a setter
+			let fields_definitions = field_helpers
+				.iter()
+				.map(|other_field| {
+					let fname = &other_field.name;
+					if other_field.order == idx {
+						if other_field.is_subsystem {
+							// We need to pack it one more time to distinguish from
+							// Fn initialization
+							quote! { #fname: OverseerFieldInit::
+								<SubsystemFieldInitMethod::<#var_type>>(
+									SubsystemFieldInitMethod::<#var_type>::Value(var)
+								)
+							}
+						}
+						else {
+							quote! { #fname: OverseerFieldInit::<#var_type>(var) }
+						}
+					}
+					else {
+						quote!{ #fname: self.#fname }
+					}
+				})
+				.collect::<Vec<_>>();
```

Comment:
Lacking some abstraction here, this is quite some visual load. Impl `trait quote::ToTokens` would solve this and encapsulate this at the very least https://docs.rs/quote/latest/quote/trait.ToTokens.html